### PR TITLE
elb_target_facts Fixes #39780

### LINKED ISSUE #39780: Find the list of all the AWS ALB target groups in which instance is registered
##### ISSUE TYPE

- Feature Idea

##### COMPONENT NAME
`elb_target`

##### ANSIBLE VERSION

```
2.5.2
```

##### SUMMARY
@wimnat I think you are the right person to ask for this feature:
For ELB, if we want to remove the instance from all the elbs, we just need to pass the instance id to the elb_instance module and again if we want to add the instance back in the same playbook, it give us a magic variable ec2_elbs, we can iterate over this variable and add back or register the instance to all the elbs which he was registered previously.

I didn't find any module, which I can use to find the list of the target groups in which the instance is register. can somebody point if know about it.

### PR DESCRIPTION
**Pull request by mostfunkyduck**
##### SUMMARY

Fixes #39780 

Adds an 'elb_target_facts' module which will gather a list of all target groups to which a given instance is registered, covering both IP and instance target groups.  This is added to a fact called 'ec2_tgs', which can be used by a playbook to change deregister/reregister nodes to the target groups.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME

lib/ansible/modules/cloud/amazon/elb_target_facts.py

##### ANSIBLE VERSION

```
ansible 2.7.0.dev0 (elb_target_facts f4d8b185f4) last updated 2018/08/01 14:26:26 (GMT -400)
  config file = None
  configured module search path = [u'/Users/ykuperman/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/ykuperman/code/ansible_devel/lib/ansible
  executable location = /Users/ykuperman/code/ansible_devel/bin/ansible
  python version = 2.7.10 (default, Jul 15 2017, 17:16:57) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.31)]

```

##### ADDITIONAL INFORMATION

```

localhost | SUCCESS => {
    "ansible_facts": {
        "ec2_tgs": [
            {
                "target_group_arn": "<arn>", 
                "target_group_type": "instance", 
                "targets": [
                    {
                        "target_az": null, 
                        "target_health": {
                            "description": "Health checks failed", 
                            "reason": "Target.FailedHealthChecks", 
                            "state": "unhealthy"
                        }, 
                        "target_id": "i-0289cbd5cc4f5f458", 
                        "target_port": 8081
                    }, 
                    {
                        "target_az": null, 
                        "target_health": {
                            "description": "Health checks failed", 
                            "reason": "Target.FailedHealthChecks", 
                            "state": "unhealthy"
                        }, 
                        "target_id": "i-0289cbd5cc4f5f458", 
                        "target_port": 8080
                    }
                ]
            }, 
            {
                "target_group_arn": "<arn>", 
                "target_group_type": "ip", 
                "targets": [
                    {
                        "target_az": "us-west-2a", 
                        "target_health": {
                            "state": "healthy"
                        }, 
                        "target_id": "172.31.22.138", 
                        "target_port": 22
                    }, 
                    {
                        "target_az": "us-west-2a", 
                        "target_health": {
                            "state": "healthy"
                        }, 
                        "target_id": "172.31.24.19", 
                        "target_port": 22
                    }
                ]
            }
        ]
    }, 
    "changed": false
}
```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/elb_target_facts.py [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+             Id={{target.target_id}},
+             Port={{target.target_port}}{%if target.target_az%},AvailabilityZone={{target.target_az}}
+             {%endif%}
+             {%endfor%}
+    with_items: "{{ec2_tgs}}"
+
+"""
+
+RETURN = """
+ec2_tgs:
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/elb_target_facts.py]: `ec2_tags`

**Comment by willthames** [comment]
Let's not use ansible_facts. Just return a dict containing instance_target_groups or something

File: lib/ansible/modules/cloud/amazon/elb_target_facts.py [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+        get_unused_target_groups={"required": False,
+                                  "default": True, "type": "bool"}
+    )
+
+    module = AnsibleAWSModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True,
+    )
+
+    if not HAS_BOTO3:
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/elb_target_facts.py]: Not needed with AnsibleAWSModule

File: lib/ansible/modules/cloud/amazon/elb_target_facts.py [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+        self.module = module
+        self.instance_id = instance_id
+        self.get_unused_target_groups = get_unused_target_groups
+        self.tgs = self._get_target_groups()
+
+    def _get_instance_ips(self):
+        """Fetch all IPs associated with this instance so that we can determine
+           whether or not an instance is in an IP-based target group"""
+        try:
+            ec2 = self.module.client(
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/elb_target_facts.py]: I really don't see the point of using a class if you're just going to create the client inside methods! 

I'd create the ec2 and elbv2 clients at `__init__` time and just use self.ec2 in the methods

File: lib/ansible/modules/cloud/amazon/elb_target_facts.py [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+except ImportError:
+    # we can handle the lack of boto3 based on the ec2 module
+    pass
+
+from ansible.module_utils.aws.core import AnsibleAWSModule
+from ansible.module_utils.ec2 import (HAS_BOTO3, camel_dict_to_snake_dict,
+                                      AWSRetry)
+
+
+class Target:
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/elb_target_facts.py]: These declarations need to be new-style classes like `Target(object):`

**Review by ryansb — CHANGES_REQUESTED** [review]
Minor style change needed, looks good otherwise.

**Review by ryansb — APPROVED** [review]

**Comment by arbabnazar** [comment]
@ryansb will it be available for 2.6 future release or 2.7.0?

**Comment by ryansb** [comment]
2.7 is the targeted release for this.

**Comment by willthames** [comment]
You can use it with 2.6 though: http://willthames.github.io/2017/12/12/using-updated-modules-with-stable-ansible.html

**Pull request merged**